### DISCUSSION TITLE
Best practices using "meta" object returned by authWithOAuth2 method

### DISCUSSION THREAD

**Discussion by maxwell-oroark** [pocketbase/pocketbase#2458] [Q&A]
Are there additional parameters that the `authWithOAuth2` method accepts so that I can automatically save certain fields returned in the `meta` object to a new user in the "users" collection?

If not, is the proper flow to just auth the user in and then patch the newly created user record with the additional fields in "meta" with a database write operation?

Additionally, I see that the there is an accessToken & refreshToken returned inside the `meta` object that I can use to interact with the provider's API. This is awesome but I am wondering why `meta` is not kept track of in the authStore class after the initial authWithOAuth2 invocation. I see it is no longer available on `pb.authStore` or `pb.authStore.onChange()` method. I guess I was wondering what the best practice is for preserving access to that accessToken when a user revives their session.

**Comment by ganigeorgiev** [comment]
>  is the proper flow to just auth the user in and then patch the newly created user record with the additional fields in "meta" with a database write operation?

Yes, if users want to persist some of the OAuth2 meta data as part of the PocketBase user, they will have to send a separate `update` request (_note that `meta.rawUser` is different for each provider, so if you have more than one OAuth2 provider enabled make sure to inspect the returned responses to adjust to your collection fields_).

>  I am wondering why meta is not kept track of in the authStore class after the initial authWithOAuth2 invocation

Because it is not part of the auth model and the meta data in general is not managed by PocketBase (we don't store the OAuth2 response on the server).
For best practices I'm not sure what to recommend as it depends on your use case. For web you can use [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) (_the same [JS SDK Security recommendation](https://github.com/pocketbase/js-sdk#security) apply_). For mobile you can use encrypted shared preferences or similar.

**Reply by maxwell-oroark** [reply]
Thanks for the quick reply! Could I piggyback of `authStore.save()` method to manage my provider tokens or would this not be advisable and I should create my own localStorage wrapper?

**Reply by ganigeorgiev** [reply]
`authStore.save()` by default accepts only the PocketBase token and the auth model as arguments, so I'm not sure how this would work. I guess it may be possible to [create your own `AuthStore`](https://github.com/pocketbase/js-sdk#authstore) (_there is a short example at the bottom_) with `save()` that would accept the `meta` as 3rd optional argument OR add manually the `meta` object as dynamic prop to the `pb.authStore.model`, but note that on `authRefresh()` call the stored `meta` may get cleared since the auth model will be replaced.

**Comment by j0sh** [comment]
Wanted to say that this is still an issue - specifically it would be nice to persist the `meta.accessToken` somewhere so it can be used after reloading the page, without the user having to re-authenticate (or refresh) for another access token.

We could store the `accessToken` ourselves within local storage, but it would be much nicer for Pocketbase to handle that on its own. 

It is also a little tricky to correctly handle the access token lifetime - see https://github.com/pocketbase/pocketbase/discussions/3617 about access token expiration. So even with an expiry available, that is another bit that needs to be stored and checked - which is more code for the user to write, and more potential for bugs. Much of that could live in the SDK, maybe checked via an `isAuthTokenValid` flag.

**Reply by ganigeorgiev** [reply]
PocketBase doesn't (and in my opinion shouldn't) care about the OAuth2 tokens and auth state.

In earlier versions they were not even part of the auth meta response and they were added later only because there were use cases where you may want to send additional requests to the provider APIs non-directly related to the PocketBase auth.

**Reply by j0sh** [reply]
> there were use cases where you may want to send additional requests to the provider APIs non-directly related to the PocketBase auth.

Yes this is exactly the use case.
